### DISCUSSION TITLE
Object Detection Datasets Overview - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#4838] [Docs]
# Object Detection Datasets Overview - Ultralytics YOLOv8 Docs

Navigate through supported dataset formats, methods to utilize them and how to add your own datasets. Get insights on porting or converting label formats.

https://docs.ultralytics.com/datasets/detect/

**Comment by s1nu5** [comment]
Hey guys, I have split my custom dataset into train, val and test. When I start training, it only indicates using the train and val data, however, I want the final accuracy measurement to be predicted with the test data. How do I do that?

**Reply by pderrenger** [reply]
@s1nu5 hello! After training your model with the train and validation datasets, you can evaluate the model's performance on your test dataset using the `val` function. You'll need to specify your test dataset in the data YAML file under the `test` key or pass the path to your test dataset directly to the `val` function. Here's a concise example:

```python
from ultralytics import YOLO

# Load your trained model
model = YOLO('path/to/your/trained_model.pt')

# Evaluate on the test dataset
results = model.val(data='path/to/your/test_data.yaml')
```

Make sure your test data YAML file is properly formatted and points to your test images and labels. This will give you the accuracy metrics on your test set. 🎯📊

**Comment by YunmeiHuanghi** [comment]
Is my custom dataset uplpaded to your platform visiable to other users? My dataset is for my research and has not realised for the public yet.

**Reply by pderrenger** [reply]
@YunmeiHuanghi no, your custom dataset is not uploaded to our platform and is not visible to other users. When you use YOLOv8, your dataset remains local on your machine or wherever you choose to host it. The training process occurs on your own hardware or cloud instances, and we do not have access to your data. Your research and datasets remain confidential unless you decide to share them. 🛡️🔒

**Comment by Ambarish-Ombrulla** [comment]
So I have created a dataset and I have trained the YOLOV8 model and I have obtained the weights. The issue is I need to train this model with another model and how do we do it. 
For example first I have trained the model on X, Y classes. Then this model I need to again train it on A, B so that I will be able to detect X , Y , A, B. 
Is this possible or do I need to train separately. Will be very helpful if you can replay as soon as possible.

**Reply by RizwanMunawar** [reply]
@Ambarish-Ombrulla, if you intend to train a model capable of detecting four classes, I highly recommend fine-tuning the model specifically for these four classes. Instead of splitting the classes into pairs of two, this approach ensures the model's ability to detect all classes effectively.

Thanks

**Comment by nadinator** [comment]
Is this YOLO format for label files any different from the YOLOv3 or YOLOv4 formats?

**Reply by RizwanMunawar** [reply]
The format for label files in Ultralytics YOLOv8 is the same as the formats used in YOLOv3 and YOLOv4.

Thanks,
Ultralytics Team!

**Comment by FyLiZo** [comment]
Hi,
I was wondering if YOLO V8 works better with loose (also occluded object area) or tight bounding boxs (only visible object area)?

**Reply by pderrenger** [reply]
@FyLiZo hello! YOLOv8, like other object detection models, generally performs better with tight bounding boxes that closely fit the visible area of the object. This allows the model to learn the precise shape and features of the objects, which is important for accurate detection, especially in crowded scenes where objects may overlap or occlude each other. However, during training, it's also beneficial to include examples of occluded objects with accurate bounding boxes to improve the model's robustness to such scenarios. 🎯🤖

**Comment by Ambarish-Ombrulla** [comment]
Thats fine but. The objects may be needed in the future so in that case
what can we do?

**Reply by RizwanMunawar** [reply]
I would suggest training three models than

- One model for the detection of the first two classes (A and B)
- One model for detection of the other two classes (C and D)
- One generic model for the detection of four classes (A, B, C, D)

**Comment by Ambarish-Ombrulla** [comment]
Thanks for the tip! . When taking models to production or any other
deployment. Inferencing on  3 models will be expensive on hardware and time
consuming?

**Reply by pderrenger** [reply]
@Ambarish-Ombrulla deploying multiple models in production can indeed be more resource-intensive and potentially slower compared to a single model, as each model requires separate computational resources for inference. However, the decision to use multiple models versus a single model should be based on the specific requirements of your application, such as the level of accuracy needed for each class and the available hardware.

If you're concerned about hardware constraints and inference time, you might consider the following strategies:
- **Model Pruning**: Reduce the size of your models by pruning less important weights.
- **Quantization**: Convert models to use lower-precision (e.g., INT8) weights and activations.
- **Model Distillation**: Train a smaller, more efficient model to mimic the performance of larger models.
- **Optimized Inference Engines**: Use optimized inference engines like TensorRT or OpenVINO to accelerate model inference on specific hardware.

Remember that YOLOv8 provides various export modes that can help optimize models for deployment. Feel free to explore the [Export](https://docs.ultralytics.com/modes/export) documentation for more details on how to prepare your models for production environments. 🚀

**Comment by ykn96** [comment]
Hey there, just to confirm
Should the annotations contain commas, or do not contains commas?

**Reply by RizwanMunawar** [reply]
@fatbringer Annotations should not contain commas, they should contain a single space after every coordinate value.

i.e

```classid x y w h```

Thanks
Ultralytics Team

**Comment by Ravina-gupt** [comment]
i got this error train: WARNING ⚠️ No labels found in C:\Users\mahaw\Desktop\Yolo8\customdataset\Thermal\Train\image.cache. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
train: New cache created: C:\Users\mahaw\Desktop\Yolo8\customdataset\Thermal\Train\image.cache
WARNING ⚠️ No labels found in C:\Users\mahaw\Desktop\Yolo8\customdataset\Thermal\Train\image.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
val: Scanning C:\Users\mahaw\Desktop\Yolo8\customdataset\Thermal\Validate\image... 0 images, 21 backgrounds, 0 corrupt: 100%|███████
val: WARNING ⚠️ C:\Users\mahaw\Desktop\Yolo8\customdataset\Thermal\Validate\image\20150711_181223_A.jpg: corrupt JPEG restored and saved

**Reply by hannaliavoshka** [reply]
Try to reorganize your dataset structure:

dataset/
|-- train/
|   |-- images/
|   |-- labels/
|-- val/
|   |-- images/
|   |-- labels/        
|-- test/
|   |-- images/
|   |-- labels/  

'images' folders should contain image files, and 'labels' folders should contain txt label files for these images.
Make sure your yaml file corresponds to this structure: train, val and test pathes should point to the train, val and test folders respectively (for example, to /train, not to /train/images).
@RizwanMunawar, please revise the documentation for the detection dataset format, it is unclear where to store label files.

**Reply by RizwanMunawar** [reply]
@Ravina-gupt @hannaliavoshka Thanks for sharing the information. we will update the docs soon!

Thanks
Ultralytics Team!

**Reply by SamaElgharabawy** [reply]
That's exactly what i'm doing and i'm still getting the same error

**Reply by pderrenger** [reply]
@SamaElgharabawy it seems like there might be an issue with the label files not being detected or not matching the expected format. Here are a few steps to troubleshoot the problem:

1. **Check Label Files**: Ensure that each image in your `train` and `val` directories has a corresponding `.txt` label file in the `labels` subdirectory. The label files should be in the same name as the image files, but with a `.txt` extension.

2. **Label Format**: Verify that the label files are in the correct YOLO format:
   ```
   <class> <x_center> <y_center> <width> <height>
   ```
   The values should be normalized between 0 and 1 relative to the image's width and height.

3. **File Paths**: Confirm that the paths in your `.yaml` file are correct and point to the directories where your images and labels are stored.

4. **No Empty Labels**: Make sure there are no empty label files. If an image does not contain any objects, it should not have a corresponding label file.

5. **Image Formats**: Check if all images are in a format supported by YOLOv8 (e.g., `.jpg`, `.png`) and that they are not corrupted. The warning about the corrupt JPEG suggests that there might be an issue with some image files.

6. **Cache File**: Delete the `image.cache` file and try running the training command again. The cache will be recreated, which can sometimes resolve issues.

7. **Dataset Configuration**: Review your dataset configuration in the `.yaml` file. It should look something like this:
   ```yaml
   train: ./customdataset/Thermal/Train/images
   val: ./customdataset/Thermal/Validate/images
   test:  # Optional path to test images
   ```

If you've checked all the above and the issue persists, please provide more details about your dataset structure and the contents of your `.yaml` file. This will help in diagnosing the problem more accurately. 🛠️

**Comment by mah92** [comment]
What considerations should we take when making a detection dataset?
1) Should the images consist of different bounding boxes or just one per image is enough? (considering that the classes come together in reality)
2) What happens if we miss drawing some bounding boxes? does it affect the training output?
3) If we have control over camera angles with respect to the object, what is a good angle spacing? Is it good to take pictures every e.g. 15 degrees in azimuth and elevation(totaling to 24 azimuth x 13 elevation = 312images per object)?
4) Is the yolovx family indifferent to inplane 2d rotation or should I make random(or systematic) 2d rotations in the dataset? 
5) What about scale? Is the algorithm indifferent or should I make random(or systematic) scales of the object in the dataset?
6) What about position? Is the algorithm, indifferent to 2d position of the object in the image(e.g. center or the border of image) or should I make random positioning?
7) What about noise/blurring/lighting? Is it indifferent to lighting or amount of noise/blurring present in the image or should I provide the closest possible conditions to the final environment?
8) What about the background? Should I provide the final environments background or just giving multiple backgrounds or a white one is enough?
9) What about occlusion? should I provide partial occlusion?

**Reply by hannaliavoshka** [reply]
Here are some basic insights about the dataset: https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/#dataset. The main idea is to keep your dataset representative of the deployed environment as much as possible, i.e. to provide the closest possible conditions to the final environment. There are some default image augmentation technics enabled during the YOLO training process, it is a good idea to try them as it is, and if the model performance is not satisfied, experiment with them. From my experience, an insignificant amount of unlabelled objects in the dataset will not hurt the model performance so much. However, it will still hurt the validation metrics, so keeping this unlabeled object amount as small as possible is a good idea. In general, it will be useful to first compile a minimal representative dataset, train it fast several times to test some conditions and parameters, see what is working best for you, and then expand the dataset and train it in this way. It will be an iterative process.

**Reply by pderrenger** [reply]
@hannaliavoshka 1) **Diversity in Bounding Boxes**: It's beneficial to have multiple bounding boxes per image, especially if classes often appear together in reality. This helps the model learn context and relationships between objects.

2) **Missing Bounding Boxes**: Missing annotations can lead to false negatives during training, which may affect the model's ability to detect those objects. It's important to label all instances of objects you wish to detect.

3) **Camera Angles**: Varied camera angles can help create a robust model. Taking pictures at systematic intervals, like every 15 degrees, can be helpful, but ensure you also include a variety of naturally occurring angles.

4) **In-Plane Rotation**: YOLO models can handle some degree of in-plane rotation, but including rotated images in the dataset can improve robustness, especially if rotations are common in the deployment environment.

5) **Scale Variations**: Including different scales of objects in your dataset is important, as it helps the model detect objects at various distances and sizes.

6) **Object Positioning**: The model should generalize to objects in various positions. Including objects near the borders can be particularly challenging and beneficial for training.

7) **Noise/Blurring/Lighting**: The model should be trained on images that reflect the conditions of the deployment environment. Including variations in noise, blur, and lighting can improve robustness.

8) **Backgrounds**: Diverse backgrounds help prevent overfitting to specific scenes. If the model will be deployed in a specific environment, make sure to include that background in the training data.

9) **Occlusion**: Including partially occluded objects can be very beneficial, as it teaches the model to recognize objects even when they're not fully visible.

Remember, the goal is to create a dataset that closely mirrors the conditions under which the model will be deployed. 🎯 For more detailed guidance, please refer to our documentation on [datasets](https://docs.ultralytics.com/datasets/detect/).

**Comment by SamaElgharabawy** [comment]
TensorBoard: Start with 'tensorboard --logdir runs/detect/train16', view at http://localhost:6006/
Freezing layer 'model.22.dfl.conv.weight'
train: Scanning /content/drive/MyDrive/detection/TRAIN/train_img/Jelly.cache... 0 images, 702 backgrounds, 0 corrupt: 100%|██████████| 702/702 [00:00<?, ?it/s]WARNING ⚠️ No labels found in /content/drive/MyDrive/detection/TRAIN/train_img/Jelly.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01), CLAHE(p=0.01, clip_limit=(1, 4.0), tile_grid_size=(8, 8))

val: Scanning /content/drive/MyDrive/detection/VALID/val_img/Jelly.cache... 0 images, 151 backgrounds, 0 corrupt: 100%|██████████| 151/151 [00:00<?, ?it/s]WARNING ⚠️ No labels found in /content/drive/MyDrive/detection/VALID/val_img/Jelly.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.

Plotting labels to runs/detect/train16/labels.jpg... 
zero-size array to reduction operation maximum which has no identity
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.00125, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
10 epochs...
Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01), CLAHE(p=0.01, clip_limit=(1, 4.0), tile_grid_size=(8, 8))

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       1/10         0G          0      110.8          0          0        640: 100%|██████████| 44/44 [10:31<00:00, 14.34s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 5/5 [00:47<00:00,  9.60s/it]                   all        151          0          0          0          0          0
WARNING ⚠️ no labels found in detect set, can not compute metrics without labels



---
#path:'/content/drive/MyDrive/detection/' # dataset root dir
train: '/content/drive/MyDrive/detection/TRAIN' # train images (relative to 'path') 4 images
val: '/content/drive/MyDrive/detection/VALID'  # val images (relative to 'path') 4 images

names:
  0: Jelly
  1: fish
  2: shark
  3: tuna




TRAIN & VALID  both have two sub-folders 1 for images and 1 for annotations , so what might be the error ?

**Reply by pderrenger** [reply]
@SamaElgharabawy it appears that the dataset is not being detected correctly during training and validation. The warning message indicates that no labels were found in the specified `.cache` files for your training and validation images. This could be due to several reasons:

1. **Incorrect Path**: The paths to your label files might be incorrect. Ensure that the paths specified in your dataset YAML file are correct and relative to the `path` variable.

2. **Missing Labels**: There might be no label files in the specified directories, or the label files might be empty.

3. **Incorrect Format**: The label files might not be in the correct format expected by YOLOv8. The labels should be in one of the supported formats, such as YOLO or COCO.

4. **Cache Files**: The `.cache` files might be outdated or corrupted. Try deleting the `.cache` files to force the dataset to be rescanned.

To resolve this issue, please check the following:

- Verify that the paths to your images and labels are correct in your dataset YAML file.
- Ensure that each image has a corresponding label file with the correct format.
- Delete the `.cache` files and rerun the training to regenerate them.
- Review the dataset formatting guidance in the Ultralytics Docs for detailed instructions on preparing your dataset.

If you continue to experience issues, you may need to manually inspect a few label files to confirm they are non-empty and formatted correctly. If the problem persists, consider revisiting the dataset preparation steps outlined in the documentation.

**Comment by Ambarish-Ombrulla** [comment]
I have the 4 pixel coordinates of the bounding box so, How do i convert to YOLOV8 dataset format

**Reply by pderrenger** [reply]
@Ambarish-Ombrulla to convert your bounding box coordinates to the YOLOv8 dataset format, you'll need to transform the coordinates from absolute pixel values to relative values with respect to the image width and height. YOLOv8 expects the bounding box in the format `[class x_center y_center width height]`, where:

- `class` is the object class integer.
- `x_center` and `y_center` are the center coordinates of the bounding box relative to the width and height of the image.
- `width` and `height` are the dimensions of the bounding box relative to the width and height of the image.

Assuming you have the top-left `(x1, y1)` and bottom-right `(x2, y2)` coordinates of the bounding box, here's how you can convert them:

1. Calculate the center coordinates `(x_center, y_center)` of the bounding box:
   ```python
   x_center = (x1 + x2) / 2
   y_center = (y1 + y2) / 2
   ```

2. Calculate the width and height of the bounding box:
   ```python
   width = x2 - x1
   height = y2 - y1
   ```

3. Normalize the center coordinates and dimensions by the image width `(img_width)` and height `(img_height)`:
   ```python
   x_center /= img_width
   y_center /= img_height
   width /= img_width
   height /= img_height
   ```

4. Format the bounding box for YOLOv8:
   ```python
   yolo_format_bbox = [class_id, x_center, y_center, width, height]
   ```

5. Save this information for each bounding box in a `.txt` file with the same name as the image file, but with a `.txt` extension. Each line in the `.txt` file corresponds to one bounding box.

For example, if you have an image `image1.jpg` with dimensions `640x480` and a bounding box with coordinates `(50, 100, 200, 300)` for class `0`, the YOLOv8 formatted bounding box would be:

```python
class_id = 0
x1, y1, x2, y2 = 50, 100, 200, 300
img_width, img_height = 640, 480

x_center = (x1 + x2) / 2 / img_width
y_center = (y1 + y2) / 2 / img_height
width = (x2 - x1) / img_width
height = (y2 - y1) / img_height

yolo_format_bbox = [class_id, x_center, y_center, width, height]
```

The corresponding line in `image1.txt` would be:
```
0 0.1953125 0.4166666666666667 0.234375 0.4166666666666667
```

Repeat this process for each bounding box in your dataset. Make sure that the `.txt` files are in the same directory as their corresponding images or in a labels folder structured correctly as per the YOLOv8 documentation.

**Comment by Salvo-Dippolito** [comment]
Is it possible to feed segmentation masks to YOLO as training data instead of the usual images and bounding box labels?

**Reply by pderrenger** [reply]
@Salvo-Dippolito yes, YOLOv8 supports training with segmentation masks. When training a model for the task of segmentation, you can provide segmentation masks along with the images. The model will learn to segment the images into different regions based on the provided masks. This is useful for tasks that require more granular output than bounding boxes, such as medical imaging or detailed scene understanding.

To train with segmentation masks, you'll need to ensure that your dataset is prepared correctly with the masks in a supported format, and you'll use a YOLOv8 model that is designed for segmentation tasks. For more details on how to set up your dataset and train a segmentation model with YOLOv8, please refer to the [Segmentation](https://docs.ultralytics.com/tasks/segment) section in the Ultralytics documentation. 📘🖌️
